group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/MultiThreadedEventProcessor.java

```diff
@@ -0,0 +1,188 @@
+     * accumulator and runs them.
+     */
+    private class EventProcessorThread extends Thread {
+        private final Logger log;
+
+        EventProcessorThread(
+            String name
+        ) {
+            super(name);
+            log = new LogContext("[" + name + "]: ").logger(EventProcessorThread.class);
+            setDaemon(false);
+        }
+
+        private void handleEvents() {
+            while (!shuttingDown) {
+                CoordinatorEvent event = accumulator.poll();
+                if (event != null) {
+                    try {
+                        log.debug("Executing event: {}.", event);
+                        event.run();
+                    } catch (Throwable t) {
+                        log.error("Failed to run event {} due to: {}.", event, t.getMessage(), t);
+                        event.complete(t);
+                    } finally {
+                        accumulator.done(event);
+                    }
+                }
+            }
+        }
+
+        private void drainEvents() {
+            CoordinatorEvent event = accumulator.poll(0, TimeUnit.MILLISECONDS);
+            while (event != null) {
+                try {
+                    log.debug("Draining event: {}.", event);
+                    event.complete(new RejectedExecutionException("EventProcessor is closed."));
+                } catch (Throwable t) {
+                    log.error("Failed to reject event {} due to: {}.", event, t.getMessage(), t);
+                } finally {
+                    accumulator.done(event);
+                }
+
+                event = accumulator.poll(0, TimeUnit.MILLISECONDS);
+            }
+        }
+
+        @Override
+        public void run() {
+            log.info("Starting");
+
+            try {
+                handleEvents();
+            } catch (Throwable t) {
+                log.error("Exiting with exception.", t);
+            }
+
+            // The accumulator is drained and all the pending events are rejected
+            // when the event processor is shutdown.
+            if (shuttingDown) {
+                log.info("Shutting down. Draining the remaining events.");
+                try {
+                    drainEvents();
+                } catch (Throwable t) {
+                    log.error("Draining threw exception.", t);
+                }
+                log.info("Shutdown completed");
+            }
+        }
+    }
+
+    /**
+     * Enqueues a new {{@link CoordinatorEvent}}.
+     *
+     * @param event The event.
+     * @throws RejectedExecutionException If the event processor is closed.
+     */
+    @Override
+    public void enqueue(CoordinatorEvent event) throws RejectedExecutionException {
+        accumulator.add(event);
+    }
+
+    /**
+     * Begins the shutdown of the event processor.
+     */
+    public synchronized void beginShutdown() {
+        if (shuttingDown) {
+            log.debug("Event processor is already shutting down.");
```

Comment:
extremely minor nit that we may not even want to change -- but in theory, we could already be done "shutting down" right? Maybe not a huge complaint, but the log could be confusing if close() has already completed.

Reply:
That's right. I think that the aim of this log is to catch cases where this class would be shutdown more than once by mistake so it does not have to 100% correct.